File: src/main/java/leetcode/permutaution/Permutation.java
package leetcode.permutaution;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Permutation {

    public static List<List<Integer>> getPermutations(Integer[] items) {
        List<List<Integer>> result = new ArrayList<>();
        calculatePermutations(result, items, new Stack<>(), new boolean[items.length]);
        return result;
    }

    private static void calculatePermutations(List<List<Integer>> result, Integer[] nums, Stack<Integer> permutation, boolean[] used) {
        if (permutation.size() == nums.length) {
            result.add(permutation.stream().toList());
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (!used[i]) {
                    used[i] = true;
                    permutation.push(nums[i]);
                    calculatePermutations(result, nums, permutation, used);
                    used[i] = false;
                    permutation.pop();
                }
            }

        }
    }

    public static void main(String[] args) {
        List<List<Integer>> permutations = getPermutations(new Integer[]{1, 2, 3});
        permutations.forEach(System.out::println);
    }
}

File: src/main/java/leetcode/permutaution/Permutation.kt
package leetcode.permutaution

import java.util.*

fun getPermutations(items: Array<Int>): List<List<Int>> {
    val result: MutableList<List<Int>> = ArrayList()
    calculatePermutations(result, items, Stack(), BooleanArray(items.size))
    return result
}

private fun calculatePermutations(
    result: MutableList<List<Int>>,
    nums: Array<Int>,
    permutation: Stack<Int>,
    used: BooleanArray
) {
    if (permutation.size == nums.size) {
        result.add(permutation.stream().toList())
    } else {
        for (i in nums.indices) {
            if (!used[i]) {
                used[i] = true
                permutation.push(nums[i])
                calculatePermutations(result, nums, permutation, used)
                used[i] = false
                permutation.pop()
            }
        }
    }
}

fun main() {
    val permutations = getPermutations(arrayOf(1, 2, 3))
    println(permutations)
}